generator/standalone.cc

```diff
@@ -0,0 +1,180 @@
+#include <standalone.h>
+#include <fstream>
+
+#include <google/protobuf/compiler/command_line_interface.h>
+#include <google/protobuf/stubs/io_win32.h>
+#include <gapic_generator.h>
+
+namespace google {
+namespace api {
+namespace codegen {
+
+namespace pb = google::protobuf;
+
+// GAPIC Standalone expects proto package as an argument. It defines the API
+// for which the GAPIC code must be generated for (i.e. to distinguish between
+// common proto descriptors and the actual API descriptors). Protobuf compiler
+// and the corresponding google::protobuf::CommandLineInterface expect the list
+// of the API-specific .proto files instead.
+//
+// This function extracts the relevant file names (those, which match the
+// package) from the set of descriptors.
+bool ExtractFileNames(
+    const std::vector<std::string>& desc_sets,
+    const std::vector<std::string>& packages,
+    std::vector<std::string>& filenames,
+    std::string* error_msg) {
+
+  for (const std::string& desc_set : desc_sets) {
+    if (desc_set.empty()) {
+      continue;
+    }
+
+    std::ifstream fstr(desc_set);
+    if (!fstr) {
+      *error_msg = "Could not open file " + desc_set;
+      return false;
+    }
+
+    pb::FileDescriptorSet bin_desc_set;
+    bool parse_result = bin_desc_set.ParseFromIstream(&fstr);
+
+    fstr.close();
```

Comment:
It is not, but it does not hurt, also allows to close the file before the processing of its content begins (otherwise it will be closed at he end of the `for` block), so we release resources as early as we stop needing them. Unless you have a strong opinion about it or it is strictly against best practices (I do not know if it is) I would prefer keeping the explicit close.

Reply:
I don't think it's too important in this case. As you pointed out, the fact that there's an explicit close method implies that it's reasonable to want to close the stream before its lexical lifetime has ended.